### PR TITLE
Tendermint sync experimental data race

### PR DESCRIPTION
**Pull request by aneequesafdar**
glienicke and some contracts cause data race, this race appeared before merging 1.9.1, so the log is from commits made before the merge:

```
==================

WARNING: DATA RACE

Write at 0x00c00181f286 by goroutine 287:

 github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).getValidators()

   /home/user98/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/engine.go:367 +0x38b

 github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).Finalize()

   /home/user98/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/engine.go:330 +0xae

 github.com/clearmatics/autonity/consensus/tendermint/core.(*core).Finalize()

   /home/user98/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/core_backend.go:44 +0x119

 github.com/clearmatics/autonity/miner.(*worker).commit()

   /home/user98/go/src/github.com/clearmatics/autonity/miner/worker.go:973 +0x53f

 github.com/clearmatics/autonity/miner.(*worker).commitNewWork()

   /home/user98/go/src/github.com/clearmatics/autonity/miner/worker.go:960 +0xe1f

 github.com/clearmatics/autonity/miner.(*worker).mainLoop()

   /home/user98/go/src/github.com/clearmatics/autonity/miner/worker.go:427 +0x128c



Previous write at 0x00c00181f286 by goroutine 542:

 github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).getValidators()

   /home/user98/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/engine.go:367 +0x38b

 github.com/clearmatics/autonity/consensus/tendermint/backend.(*Backend).Finalize()

   /home/user98/go/src/github.com/clearmatics/autonity/consensus/tendermint/backend/engine.go:330 +0xae

 github.com/clearmatics/autonity/consensus/tendermint/core.(*core).Finalize()

   /home/user98/go/src/github.com/clearmatics/autonity/consensus/tendermint/core/core_backend.go:44 +0x119

 github.com/clearmatics/autonity/core.(*StateProcessor).Process()

   /home/user98/go/src/github.com/clearmatics/autonity/core/state_processor.go:79 +0x92a

 github.com/clearmatics/autonity/core.(*BlockChain).insertChain()

   /home/user98/go/src/github.com/clearmatics/autonity/core/blockchain.go:1212 +0xce2

 github.com/clearmatics/autonity/core.(*BlockChain).InsertChain()

   /home/user98/go/src/github.com/clearmatics/autonity/core/blockchain.go:1100 +0x1382

 github.com/clearmatics/autonity/eth.NewProtocolManager.func6()

   /home/user98/go/src/github.com/clearmatics/autonity/eth/handler.go:207 +0x50c

 github.com/clearmatics/autonity/eth/fetcher.(*Fetcher).insert.func1()

   /home/user98/go/src/github.com/clearmatics/autonity/eth/fetcher/fetcher.go:669 +0x7dd



Goroutine 287 (running) created at:

 github.com/clearmatics/autonity/miner.newWorker()

   /home/user98/go/src/github.com/clearmatics/autonity/miner/worker.go:219 +0x9d0

 github.com/clearmatics/autonity/miner.New()

   /home/user98/go/src/github.com/clearmatics/autonity/miner/miner.go:61 +0xe8

 github.com/clearmatics/autonity/eth.New()

   /home/user98/go/src/github.com/clearmatics/autonity/eth/backend.go:215 +0x1992

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:458 +0x2b3

 fmt.Fscanf()

   /usr/local/go/src/fmt/scan.go:143 +0xcf

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

   /usr/local/go/src/fmt/scan.go:114 +0x193

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:458 +0x2b3

 fmt.Fscanf()

   /usr/local/go/src/fmt/scan.go:143 +0xcf

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

   /usr/local/go/src/fmt/scan.go:114 +0x193

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:458 +0x2b3

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb.(*DB).checkAndCleanFiles()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/db_util.go:52 +0x2e1

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb.openDB()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/db.go:130 +0x9f9

 fmt.(*ss).doScanf()

   /usr/local/go/src/fmt/scan.go:1193 +0x377

 fmt.Fscanf()

   /usr/local/go/src/fmt/scan.go:143 +0xcf

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

   /usr/local/go/src/fmt/scan.go:114 +0x193

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:458 +0x2b3

 fmt.Fscanf()

   /usr/local/go/src/fmt/scan.go:143 +0xcf

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

   /usr/local/go/src/fmt/scan.go:114 +0x193

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:458 +0x2b3

 fmt.Fscanf()

   /usr/local/go/src/fmt/scan.go:143 +0xcf

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

   /usr/local/go/src/fmt/scan.go:114 +0x193

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:458 +0x2b3

 fmt.Fscanf()

   /usr/local/go/src/fmt/scan.go:143 +0xcf

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

   /usr/local/go/src/fmt/scan.go:114 +0x193

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:458 +0x2b3

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb.(*DB).recoverJournal()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/db.go:476 +0xc5

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb.openDB()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/db.go:125 +0x9d5

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb.Open()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/db.go:197 +0x1b0

 fmt.Fscanf()

   /usr/local/go/src/fmt/scan.go:143 +0xcf

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.fsParseName()

   /usr/local/go/src/fmt/scan.go:114 +0x193

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage.(*fileStorage).List()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:458 +0x2b3

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb.(*session).recover.func1()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/session.go:113 +0xf4

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb.(*session).recover()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/session.go:121 +0x190b

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb.Open()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/db.go:183 +0xda

 github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb.OpenFile()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/github.com/syndtr/goleveldb/leveldb/db.go:219 +0xa6

 github.com/clearmatics/autonity/ethdb.NewLDBDatabase()

   /home/user98/go/src/github.com/clearmatics/autonity/ethdb/database.go:76 +0x39c

 github.com/clearmatics/autonity/node.(*ServiceContext).OpenDatabase()

   /home/user98/go/src/github.com/clearmatics/autonity/node/service.go:47 +0x1b2

 github.com/clearmatics/autonity/eth.CreateDB()

   /home/user98/go/src/github.com/clearmatics/autonity/eth/backend.go:247 +0xa9

 github.com/clearmatics/autonity/eth.New()

   /home/user98/go/src/github.com/clearmatics/autonity/eth/backend.go:126 +0x1be

 github.com/clearmatics/autonity/consensus/test.makeValidator.func1()

   /home/user98/go/src/github.com/clearmatics/autonity/consensus/test/test_helpers.go:150 +0x3b6

 github.com/clearmatics/autonity/node.(*Node).Start()

   /home/user98/go/src/github.com/clearmatics/autonity/node/node.go:182 +0x723

 github.com/clearmatics/autonity/consensus/test.(*testNode).startNode()

   /home/user98/go/src/github.com/clearmatics/autonity/consensus/test/tendermint_test.go:656 +0x500

 github.com/clearmatics/autonity/consensus/test.runTest.func1()

   /home/user98/go/src/github.com/clearmatics/autonity/consensus/test/tendermint_test.go:560 +0x41

 github.com/clearmatics/autonity/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1()

   /home/user98/go/src/github.com/clearmatics/autonity/vendor/golang.org/x/sync/errgroup/errgroup.go:57 +0x64



Goroutine 542 (running) created at:

 github.com/clearmatics/autonity/eth/fetcher.(*Fetcher).insert()

   /home/user98/go/src/github.com/clearmatics/autonity/eth/fetcher/fetcher.go:643 +0x309

 github.com/clearmatics/autonity/eth/fetcher.(*Fetcher).loop()

   /home/user98/go/src/github.com/clearmatics/autonity/eth/fetcher/fetcher.go:313 +0x854

==================
```

### CODE REVIEW

**Comment by JekaMas** [comment]
Unfortunately, it's a part of more general problem that autonity starts multiple times and concurrently.

**Comment by aneequesafdar** [comment]
So what do you recommend to do?

File: consensus/tendermint/backend/engine.go [hunk 1]
```diff
@@ -380,25 +380,27 @@ func (sb *Backend) getValidators(header *types.Header, chain consensus.ChainRead
...
 		validators, err = sb.retrieveSavedValidators(1, chain)
 		if err != nil {
 			return nil, err
 		}
 
 		// Deploy Glienicke network-permissioning contract
-		_, sb.glienickeContract, err = sb.blockchain.DeployGlienickeContract(state, header)
+		var glienickeAddr common.Address
+		_, glienickeAddr, err = sb.blockchain.DeployGlienickeContract(state, header)
+		sb.setGlienickeContract(glienickeAddr)
```
Comment (JekaMas) [inline consensus/tendermint/backend/engine.go]: It sets a new address even if error happened.

File: consensus/tendermint/core/prevote.go [hunk 2]
```diff
@@ -70,28 +68,17 @@ func (c *core) handlePrevote(ctx context.Context, msg *Message) error {
...
-					"proposalHeight", preVote.Height.Uint64(),
-					"proposalRound", preVote.Round.Uint64(),
-					"currentHeight", c.currentRoundState.height.Uint64(),
-					"currentRound", c.currentRoundState.round,
-				)
-
-				return fmt.Errorf("handlePrevote. vote height %d, round %d, current height, %d round %d",
-					preVote.Height.Uint64(), preVote.Round.Uint64(),
-					c.currentRoundState.height.Uint64(), c.currentRoundState.round,
+				oldRoundState = *NewRoundState(
```
Comment (JekaMas) [inline consensus/tendermint/core/prevote.go]: Is it always non-nil value?

Comment (aneequesafdar) [reply consensus/tendermint/core/prevote.go]: NewRoundState() return a `&roundstate{...}`, so yeah it should be non-nil

File: consensus/tendermint/backend/engine.go (hunk 1 above)
Comment (aneequesafdar) [reply consensus/tendermint/backend/engine.go]: yeah you are correct I will change it

**Pull request merged**